src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Remote.cs

```diff
@@ -281,41 +281,320 @@ public async Task MergePullRequestAsync(string pullRequestUrl, MergePullRequestP
+                if (!nodeCache.TryGetValue($"{currentDependency.RepoUri}@{currentDependency.Commit}", out rootNode))
+                {
+                    DependencyGraph dependencyGraph =
+                        await DependencyGraph.BuildRemoteDependencyGraphAsync(remoteFactory,
+                                                                              null,
+                                                                              currentDependency.RepoUri,
+                                                                              currentDependency.Commit,
+                                                                              dependencyGraphBuildOptions,
+                                                                              _logger);
+                    // Cache all nodes in this built graph.
+                    foreach (DependencyGraphNode node in dependencyGraph.Nodes)
+                    {
+                        if (!nodeCache.ContainsKey($"{node.RepoUri}@{node.Commit}"))
+                        {
+                            nodeCache.Add($"{node.RepoUri}@{node.Commit}", node);
+                        }
+                    }
+                    rootNode = dependencyGraph.Root;
+                }
+
+                // Now do the lookup to find the element in the tree for each item in the update list
+                foreach (DependencyDetail dependencyInUpdateChain in updateList)
+                {
+                    (Asset coherentAsset, Build buildForAsset) =
+                        FindAssetInBuildTree(dependencyInUpdateChain.Name, rootNode);
+
+                    if (coherentAsset == null)
+                    {
+                        // This is an invalid state. We can't satisfy the
+                        // constraints so they should either be removed or pinned.
+                        throw new DarcException($"Unable to update {dependencyInUpdateChain.Name} to have coherency with " +
+                            $"parent {dependencyInUpdateChain.CoherentParentDependencyName}. No matching asset found in tree. " +
+                            $"Either remove the coherency attribute or mark as pinned.");
+                    }
+
+                    string buildRepoUri = buildForAsset.AzureDevOpsRepository ?? buildForAsset.GitHubRepository;
 
-        public async Task<List<DependencyDetail>> GetRequiredUpdatesAsync(
+                    if (dependencyInUpdateChain.Name == coherentAsset.Name &&
+                        dependencyInUpdateChain.Version == coherentAsset.Version &&
+                        dependencyInUpdateChain.Commit == buildForAsset.Commit &&
+                        dependencyInUpdateChain.RepoUri == buildRepoUri)
+                    {
+                        continue;
+                    }
+
+                    DependencyDetail updatedDependency = new DependencyDetail(dependencyInUpdateChain);
+                    updatedDependency.Name = coherentAsset.Name;
```

Comment:
You could use the 
```
{
    Name = coherentAsset.Name,
}
```
style here to avoid repeating the updatedDependency over and over (and make it clear to this poor reviewer that this is just setting a bunch of properties on the object we just created)